File: leetcode/90.subsetswithdup/ts/subsetswithdup.ts
const subsetsWithDup = (nums: number[]): number[][] => {
    const res: number[][] = []
    const path: number[] = []

    const backtrack = (start: number) => {
        res.push([...path])
        if (start === nums.length) return

        for (let i = start; i < nums.length; i++) {
            if (i > start && nums[i] === nums[i - 1]) continue

            path.push(nums[i])
            backtrack(i + 1)
            path.pop()
        }
    }

    nums.sort((a, b) => a - b)
    backtrack(0)
    return res
}

File: leetcode/90.subsetswithdup/go/subsetswithdup.go
package subsetswithdup

import (
	"slices"
)

func subsetsWithDup(nums []int) [][]int {
	res := make([][]int, 0)
	path := make([]int, 0)

	var backtrack func(int)

	backtrack = func(start int) {
		tmp := make([]int, len(path))
		copy(tmp, path)
		res = append(res, tmp)

		if len(path) == len(nums) {
			return
		}

		for i := start; i < len(nums); i++ {
			if i > start && nums[i] == nums[i-1] {
				continue
			}

			path = append(path, nums[i])
			backtrack(i + 1)
			path = path[:len(path)-1]
		}
	}

	slices.Sort(nums)
	backtrack(0)
	return res
}
